docs: test replay docs (#5349)
cypress-io/cypress-documentation: docs/guides/cloud/debugging/test-replay.mdx

---
title: Test Replay
sidebar_position: 20
---

:::info

## <Icon name="graduation-cap" /> What you'll learn

- What Test Replay is and why you should use it
- How to get started with Test Replay in Cypress Cloud
- How to debug failed and flaky CI test runs more accurately and efficiently

:::

> 

Reviewer: should probably put an info tip around here to let users know its cy13 and up only

## What Is Test Replay?

Test Replay in Cypress Cloud allows developers to accurately and efficiently debug failed and flaky CI test runs. It captures every detail of test runs, enabling developers to replay and interact with the DOM, inspect network events, and console events as they happened during CI. This eliminates the need to replicate failed tests locally and enhances the ability to debug complex applications with confidence and speed.

> ## Why Test Replay?

Reviewer: wonder if a better title would be "Test Replay vs Screenshots and Videos"

Author: I can get down with that, I'll include as an h3 in this section.

## Getting Started

:::info

&#8239<Icon name="check-circle" /> Test Replay is available to **all** Cypress Cloud plans at no additional cost, subject to usage limits. Free plans allow up to 500 test results per month, while exceeding this limit requires upgrading the plan.

:::

Test Replay is automatically enabled when updating to `v13` of Cypress and an opt-out is provided through project-level settings. Learn about what this means for capturing video of test failure in Cypress Cloud in the [See also](#See-also) section below.

> All you need to do after recording test runs is log in to your Cypress Cloud account and leverage these superior debugging capabilities. No additional software, configuration or changes to the existing test recording process is required. For example, you do **_not_** need to edit your [Cypress configuration](/guides/references/configuration) file or pass any flags or new commands when recording via [the command line](/guides/guides/command-line) in order to enable or use Test Replay.

Reviewer: superior feels a tad strong here. Maybe "additional" instead?

Author: With the video default changes in v13, I think it feels less correct to say "additional", considering the out of the box config. "Improved" sounds underwhelming. Maybe just "new". I'll play around.

<DocsImage
  src="/img/guides/cloud/test-replay/opt-out.png"
  alt="Test Replay opt-out controls"
/>

## Troubleshooting

**If Test Replay is not present or disabled in Cypress Cloud:**

1. Confirm runs have been recorded to Cypress Cloud using Cypress `v13` or later.
2. Ensure tests were recorded using a Chromium-based browser(Chrome, Edge, Electron).
3. Check that Test Replay is enabled in project settings.

**If performance loss is experienced in Cypress with Test Replay enabled:**

Cypress may become less performant under certain circumstances with Test Replay. You may be able to balance performance by taking some additional measures.

Cypress uses a powerful [command log](/guides/core-concepts/cypress-app#Command-Log) locally to display every command executed in your spec. Yet, the command log in the test replay on Cypress Cloud is recreated at a _later_ time. If you are not using screenshots or video or don't need the command log displayed in these artifacts, you can [disable the command log](/guides/references/troubleshooting#Disable-the-Command-Log) when running the test.

> This can positively affect test speed without loosing the interactive benefit of the command log during the Test Replay. You can still keep the command log enabled for test runs and artifacts in non-Chromium browsers.

Reviewer 2: we made a decision to disable the command log by default - we need to update docs for that and update this section @jaffrepaul cc @jennifer-shehane

---
title: Test Replay
sidebar_position: 20
---

:::info

## <Icon name="graduation-cap" /> What you'll learn

- What Test Replay is and why you should use it
- How to get started with Test Replay in Cypress Cloud
- How to debug failed and flaky CI test runs more accurately and efficiently

:::

:::caution

> {/* todo: change changelog url to real page when it's ready/won't break build */} <br/>

Reviewer 3: ```suggestion
```

> &#8239;<Icon name="exclamation-triangle" /> Test Replay is available when recording tests to Cypress Cloud using Cypress `v13` and up. See the [Cypress v13 changelog](/guides/references/changelog#12-0-0).

Reviewer 3: would this actually break the build?
```suggestion
&#8239;<Icon name="exclamation-triangle" /> Test Replay is available when recording tests to Cypress Cloud using Cypress `v13` and up. See the [Cypress v13 changelog](/guides/references/changelog#13-0-0).
```

Reviewer 3: OK that makes sense I guess we can resolve once its in the v13 branch

> Test Replay currently supports Chromium-based browsers(Chrome, Edge, Electron) only.

Reviewer 3: ```suggestion
Test Replay currently supports Chromium-based browsers (Chrome, Edge, Electron) only.
```

:::

## What Is Test Replay?

Test Replay in Cypress Cloud allows developers to accurately and efficiently debug failed and flaky [continuous integration](/guides/continuous-integration/introduction) test runs. It captures every test run detail, enabling developers to replay it and inspect the DOM, network requests, console logs, JavaScript errors, and element rendering as they happened during CI.

### Test Replay vs Screenshots and Videos

Before Cypress `v13`, test failures in CI have historically been captured through screenshots, videos, and stack trace outputs, but these artifacts provide limited information. Developers often need more context to understand the root cause of failures, requiring them to reproduce the issue locally, which may not always be possible due to environmental or network conditions. This knowledge gap between test failures and debugging capabilities leads to increased debugging time and potentially deploying unresolved bugs to production.

### How Does Test Replay Advance Your Workflow?

> Test Replay complements the existing visual artifacts provided by Cypress Cloud. It eliminates the need for local test replication by enabling developers to time-travel to the exact moment of test failure. Developers gain access to previously inaccessible details and feedback loops, enabling faster and more accurate debugging, and instilling confidence in handling test failures in CI. Replacing video recordings with Test Replay provides a superior debugging experience that is sharable and supports team collaboration with efficient workflows.

Author: @jennifer-shehane which part exactly would be covered a gif or video?

> Test Replay complements the existing visual artifacts provided by Cypress Cloud. It eliminates the need for local test replication by enabling developers to time-travel to the exact moment of test failure. Developers gain access to previously inaccessible details and feedback loops, enabling faster and more accurate debugging, and instilling confidence in handling test failures in CI. Replacing video recordings with Test Replay provides a superior debugging experience that is sharable and supports team collaboration with efficient workflows.

Reviewer 4: It's just a sale thing really. Test Replay is so visual, it's a quick way to visually see what we're talking about instead of reading a wall of text explaining it.

## Getting Started

:::info

&#8239<Icon name="check-circle" /> Test Replay is available to **all** Cypress Cloud plans at no additional cost, subject to usage limits.

> Free plans allow up to 500 test results per month, while exceeding this limit requires upgrading the plan.

Author: Yeah, these references has been discussed a lot recently as details have changed over time. 

It was a request of @pstakoun to add it here. If we feel it's good to keep, I can look into creating a partial so that any refs will be powered by one piece of markdown.

Please see our [Pricing page](https://www.cypress.io/pricing) for more details.

:::

> Test Replay is automatically enabled when updating to `v13` of Cypress and an opt-out is provided through project-level settings. Learn about what this means for capturing video of test failure in Cypress Cloud under ["See also"](#See-also) below.

Reviewer 3: Not necessary, but a sentence break might be good here
```suggestion
Test Replay is automatically enabled when updating to `v13` of Cypress. If desired, an opt-out is provided through project-level settings. Learn about what this means for capturing video of test failure in Cypress Cloud under ["See also"](#See-also) below.
```

  src="/img/guides/cloud/test-replay/test-replay-header.jpg"
  alt="Test Replay header"
/>

### Command Log

The left side of the screen displays the [command log](/guides/core-concepts/cypress-app#Command-Log) just like in the Cypress app, where you can step through, inspect and [time travel](/guides/core-concepts/cypress-app#Time-traveling) debug your test.

<DocsImage
  src="/img/guides/core-concepts/cypress-app/command-log.png"
  alt="Test Replay Developer Tools panel"
  width="60%"
/>

### Developer Tools

Dive into network-level traffic, console events and inspect the application under test just as you can in the browser.

<DocsImage
  src="/img/guides/cloud/test-replay/dev-tools.png"
  alt="Test Replay Developer Tools panel"
  width="60%"
/>

### Replay Controls

Toggle each test attempt, play back errors at a variety of speeds or scrub through the captured error by dragging the playhead along the timeline.

<DocsImage
  src="/img/guides/cloud/test-replay/scrubber.png"
  alt="Test Replay scrubber controls"
/>

## Opt out of Test Replay

:::warning

<Icon name="exclamation-triangle" /> Do not use sensitive data in your tests. This
will be visible in the test replay.

> :::

Author: This was also a request to add from @pstakoun. I'll shift to the masking effort likely next week and we can include there if we're all in agreement. 

cc @jennifer-shehane

> :::

Reviewer 4: @jaffrepaul 

```
Please refer to the [Cypress Cloud Terms of Use](https://cloud.cypress.io/terms-of-use) and our [Security & Compliance](https://www.cypress.io/security/) guide for specifics around reasonable use of Cypress Cloud, data storage and security at Cypress.
```

Also maybe another section around
```
Test Replays, including the test data leveraged within your tests, are visible to everyone with access to your project.
```

> :::

Author: added in https://github.com/cypress-io/cypress-documentation/pull/5349/commits/b2cad3cae5955793c209e0c37440d77927a26cea

If you do not want to capture test data for replay and debugging purposes simply "shut off" Test Replay in project settings.

<DocsImage
  src="/img/guides/cloud/test-replay/opt-out.png"
  alt="Test Replay opt-out controls"
/>

## Troubleshooting

**If Test Replay is not present or disabled in Cypress Cloud:**

1. Confirm runs have been recorded to Cypress Cloud using Cypress `v13` or later.

> 2. Ensure tests were recorded using a Chromium-based browser(Chrome, Edge, Electron).

Reviewer 3: ```suggestion
2. Ensure tests were recorded using a Chromium-based browser (Chrome, Edge, Electron).
```

3. Check that Test Replay is enabled in Cypress Cloud project settings.

## Command Log Notes

Cypress uses a powerful [command log](/guides/core-concepts/cypress-app#Command-Log) locally to display every command executed in your spec. The Cypress command log is visible during `cypress open` mode, but is also rendered ([headless](/guides/guides/command-line#Options)-ly) during `cypress run` so that it displays within captured screenshots or videos, to assist in debugging CI failures.

### Problem

Rendering the Cypress command log while running tests via `cypress run` often has performance implications, resulting in a slower test runtime, especially on lower resourced machines.

### Solution

With Test Replay, Cypress regenerates the entire UI within the Cloud for viewing the run and the debugging CI failures. Additionally, displaying the Cypress command log in videos and screenshots is less valuable for users using Test Replay.

> Because of this, as of Cyrpess `v13`, the Cypress command log is no longer rendered during `cypress run` for users using Test Replay and `video: false`, with an option to opt out of this behavior.

Reviewer 3: Have we made the decision to no longer show the reporter by default? This was up in the air last time I checked and not sure if a decision was made.

Reviewer 3: Sounds like the decision to hide it is made. we might want to reference the new CLI options here in case users want to use them (original comment https://github.com/cypress-io/cypress-documentation/pull/5349#discussion_r1297296350)

> Because of this, as of Cyrpess `v13`, the Cypress command log is no longer rendered during `cypress run` for users using Test Replay and `video: false`, with an option to opt out of this behavior.

Reviewer 3: I just noticed Cypress is spelt wrong as well

> Because of this, as of Cyrpess `v13`, the Cypress command log is no longer rendered during `cypress run` for users using Test Replay and `video: false`, with an option to opt out of this behavior.

Author: Gotcha, I pulled this nearly directly from clickup. Is that logic being added at some point or not any longer? @jennifer-shehane

> Because of this, as of Cyrpess `v13`, the Cypress command log is no longer rendered during `cypress run` for users using Test Replay and `video: false`, with an option to opt out of this behavior.

Reviewer 4: @jaffrepaul This feature is described in this docs PR: https://github.com/cypress-io/cypress-documentation/pull/5424

When `video: true` is set, the Cypress command log will be captured as part of that video recording. You can still keep the command log enabled for test runs and artifacts in non-Chromium browsers.

## See also

> // TODO: replace urls for v13 branch merge

Reviewer 3: ```suggestion
```

// changelog- /guides/references/changelog#13-0-0

- [Test Replay FAQ](/faq/questions/cloud-faq#Test-Replay)

> - [Cypress v13 changelog](/guides/references/changelog#12-0-0)

Reviewer 3: ```suggestion
- [Cypress v13 changelog](/guides/references/changelog#13-0-0)
```

---
title: Test Replay
sidebar_position: 20
---

:::info

## <Icon name="graduation-cap" /> What you'll learn

- What Test Replay is and why you should use it
- How to get started with Test Replay in Cypress Cloud
- How to debug failed and flaky CI test runs more accurately and efficiently

:::

:::caution

> {/* todo: change changelog url to real page when it's ready/won't break build */} <br/>

Reviewer 3: Is the goal to remove this once it's in the v13 branch?

Reviewer 3: nevermind saw the comment on the resolution. going to resolve

## Accessing Test Replay

Accessing the replay of your test is as simple as reviewing any other test artifact in Cypress Cloud.

1. Hover on a [test for review](/guides/cloud/debugging/recorded-runs#Tests-for-Review) under the Latest Runs Overview.

   <DocsImage
     src="/img/guides/cloud/test-replay/replay_row_magnified.jpg"
     alt="Run overview row"
   />

2. Or navigate to the [test detail sidebar](/guides/cloud/debugging/recorded-runs#Test-detail-sidebar) header or above each attempt in the errors panel. The header option will replay the first failing test while the option above each attempt will replay the affiliated test.

   <DocsImage
     src="/img/guides/cloud/test-replay/replay_sidebar_magnified.jpg"
     alt="Run sidebar"
     width="60%"
   />

## Debugging With Test Replay

Both options above display the Test Replay interface, which we'll unpack from top to bottom in more detail below.

<DocsImage
  src="/img/guides/cloud/test-replay/replay-demo.gif"
  alt="Test Replay Demo"
/>

### Run header

> The left side of the header displays the Cypress Cloud test run ID #, with the Git branch name along with the spec, platform and browser information. Hovering on this exposes additional test meta data.

Reviewer 3: I read this weird the first time and wondering if this helps with readability? 
```suggestion
The left side of the header displays the Cypress Cloud test run ID number (#), with the Git branch name along with the spec, platform and browser information. Hovering on this exposes additional test meta data.
```

<DocsImage
  src="/img/guides/cloud/test-replay/test-replay-header.jpg"
  alt="Test Replay header"
/>

### Command Log

The left side of the screen displays the [command log](/guides/core-concepts/cypress-app#Command-Log) just like in the Cypress app, where you can step through, inspect and [time travel](/guides/core-concepts/cypress-app#Time-traveling) debug your test.

<DocsImage
  src="/img/guides/cloud/test-replay/command-log.png"
  alt="Test Replay Developer Tools panel"
  width="60%"
/>

### Developer Tools

Dive into network-level traffic, console events and inspect the application under test just as you can in the browser.

<DocsImage
  src="/img/guides/cloud/test-replay/dev-tools.png"
  alt="Test Replay Developer Tools panel"
  width="60%"
/>

### Replay Controls

Toggle each test attempt, play back errors at a variety of speeds or scrub through the captured error by dragging the playhead along the timeline.

<DocsImage
  src="/img/guides/cloud/test-replay/scrubber.png"
  alt="Test Replay scrubber controls"
/>

> ## Opt out of Test Replay

Reviewer 3: We say `Opt-in` for the video config changes in the `MajorVersionWelcome` component: https://github.com/cypress-io/cypress/pull/27549/files#diff-e1690da5e8832a5783a622b238019fa8a00d03899e74bc558b08557ded5c4593R86. I wonder if we should use the same spelling here?
```suggestion
## Opt-out of Test Replay
```

:::info

Please refer to the [Cypress Cloud Terms of Use](https://cloud.cypress.io/terms-of-use) and our [Security & Compliance](https://www.cypress.io/security/) guide for specifics around reasonable use of Cypress Cloud, data storage and security at Cypress.

Test Replays, including the test data leveraged within your tests, are visible to everyone with access to your project.

:::

> If you do not want to capture test data for replay and debugging purposes simply "shut off" Test Replay in project settings.

Reviewer 3: ```suggestion
If you do not want to capture test data for replay and debugging purposes, simply "shut off" Test Replay in project settings.
```

3. Check that Test Replay is enabled in Cypress Cloud project settings.
4. Review the standard output of the test run to ensure there was not an error capturing Test Replay. If you encounter an error capturing Test Replay, please open an issue so that we can investigate.

## Command Log Notes

Cypress uses a powerful [command log](/guides/core-concepts/cypress-app#Command-Log) locally to display every command executed in your spec. The Cypress command log is visible during `cypress open` mode, but is also rendered ([headless](/guides/guides/command-line#Options)-ly) during `cypress run` so that it displays within captured screenshots or videos, to assist in debugging CI failures.

### Problem

Rendering the Cypress command log while running tests via `cypress run` often has performance implications, resulting in a slower test runtime, especially on lower resourced machines.

### Solution

With Test Replay, Cypress regenerates the entire UI within the Cloud for viewing the run and the debugging CI failures. Additionally, displaying the Cypress command log in videos and screenshots is less valuable for users using Test Replay.

> Because of this, as of Cypress `v13`, the Cypress command log is no longer rendered during `cypress run` for users using Test Replay, with an option to opt out of this behavior.

Reviewer 3: ```suggestion
Because of this, as of Cypress `v13`, the Cypress command log is no longer rendered during `cypress run` for users using Test Replay, with an option to opt-out of this behavior.
```

sidebar_position: 20
---

:::info

## <Icon name="graduation-cap" /> What you'll learn

- What Test Replay is and why you should use it
- How to get started with Test Replay in Cypress Cloud
- How to debug failed and flaky CI test runs more accurately and efficiently

:::

:::caution

{/* todo: change changelog url to real page when it's ready/won't break build */} <br/>
&#8239;<Icon name="exclamation-triangle" /> Test Replay is available when recording tests to Cypress Cloud using Cypress `v13` and up. See the [Cypress v13 changelog](/guides/references/changelog#12-0-0).

:::

:::caution

> Test Replay currently supports Chromium-based browsers only.

Reviewer 2: I wonder if we should elaborate on what browsers are / are not supported – there may be customers who do not know what browsers are Chromium-based.

Author: Can't hurt 👍

### How Does Test Replay Advance Your Workflow?

Test Replay complements the existing visual artifacts provided by Cypress Cloud by allowing developers to directly interact with tests as they occurred in CI. It eliminates the need for local test replication by enabling developers to time-travel to the exact moment of test failure, replaying and interacting with network requests, console logs, JavaScript errors, and element rendering. With Test Replay, developers gain access to previously inaccessible details and feedback loops, enabling faster and more accurate debugging, and instilling confidence in handling test failures in CI. Replacing video recordings with Test Replay provides a superior debugging experience.

## Getting Started

:::tip

> &#8239<Icon name="check-circle" /> Test Replay is available to **all** Cypress Cloud plans at no additional cost, subject to usage limits. Free plans allow up to 500 test results per month, while exceeding this limit requires upgrading the plan.

Reviewer 2: Maybe link out to pricing or something here – not sure what pattern we use elsewhere

Author: No real consistent pattern for this. I'll add the basic line from the FAQs
